### PR TITLE
Akka.Discovery

### PR DESCRIPTION
**Pull request by ismaelhamed**
Akka Discovery provides an interface around various ways of locating services. The built in methods are:

- [x] Configuration
- [x] Aggregate
- [ ] DNS

**Current known limitation**

Endpoints should be defined as a list of configs:

```
akka.discovery.config.services = {
  service1 = {
    endpoints = [
      {
        host = "cat"
        port = 1233
      },
      {
        host = "dog"
        port = 1234
      }
    ]
  }
}
```

But AFAIK this is not possible with current Hocon implementation, at least unless [#293](https://github.com/akkadotnet/HOCON/issues/293) is implemented. So, in the mean time I have resorted to define endpoints as a lists of strings (much like we do with `seed-nodes`):

```
akka.discovery.config.services = {
  service1 = {
    endpoints = [
        "cat:1233",
        "dog:1234"
    ]
  }
}
```

### CODE REVIEW

File: src/core/Akka.Discovery.Tests/Akka.Discovery.Tests.csproj
Comment (ismaelhamed) [inline src/core/Akka.Discovery.Tests/Akka.Discovery.Tests.csproj:8]: I  hope this is fine for new proyects.

File: src/core/Akka.Discovery/Aggregate/AggregateServiceDiscovery.cs
Comment (ismaelhamed) [inline src/core/Akka.Discovery/Aggregate/AggregateServiceDiscovery.cs:57]: `Resolve` is recursive. I think its fine, but it's open for improvents.

File: src/core/Akka.Discovery/Akka.Discovery.csproj
Comment (ismaelhamed) [inline src/core/Akka.Discovery/Akka.Discovery.csproj:10]: I hope this is fine for new projects.

File: src/core/Akka.Discovery/Config/ConfigServiceDiscovery.cs
Comment (ismaelhamed) [inline src/core/Akka.Discovery/Config/ConfigServiceDiscovery.cs:28]: Missing `GetConfigList`, not currently supported

File: src/core/Akka.Discovery.Tests/Akka.Discovery.Tests.csproj
Comment (Aaronontheweb) [reply src/core/Akka.Discovery.Tests/Akka.Discovery.Tests.csproj:8]: Yep, that should be fine as long as we don't use default interfaces - which aren't binary backwards compatible

File: src/core/Akka.Discovery/Config/ConfigServiceDiscovery.cs
Comment (Aaronontheweb) [reply src/core/Akka.Discovery/Config/ConfigServiceDiscovery.cs:28]: @Arkatufus is this something we need to add to the HOCON spec?

**Comment by Aaronontheweb** [comment]
@ismaelhamed looks like you're off to a great start! I'll do a deeper review soon - but do we need to wait until the DNS discovery is available before we consider merging this? Just want to know if I should wait for that before diving in.

File: src/core/Akka.Discovery/Config/ConfigServiceDiscovery.cs
Comment (ismaelhamed) [reply src/core/Akka.Discovery/Config/ConfigServiceDiscovery.cs:28]: I've linked a issue on the HOCON repo to this

**Comment by ismaelhamed** [comment]
Since both `Config` and `Aggregate` options are ready, IMO you should merge this as soon as you approve it. There seems to be some work to do on the `Akka.Io.Dns` in order to implement the DNS API, and that's something that you guys might want to plan carefully.

**Comment by Aaronontheweb** [comment]
Going to review this today

File: src/core/Akka.Discovery/Aggregate/AggregateServiceDiscovery.cs
Comment (Aaronontheweb) [reply src/core/Akka.Discovery/Aggregate/AggregateServiceDiscovery.cs:57]: if someone runs into an issue with it, they we can always change it later

File: src/core/Akka.Discovery.Tests/Aggregate/AggregateServiceDiscoverySpec.cs
Comment (Aaronontheweb) [inline src/core/Akka.Discovery.Tests/Aggregate/AggregateServiceDiscoverySpec.cs:2]: File name is technically wrong here

File: src/core/Akka.Discovery/Aggregate/AggregateServiceDiscovery.cs
Comment (Aaronontheweb) [inline src/core/Akka.Discovery/Aggregate/AggregateServiceDiscovery.cs:37]: LGTM

File: src/core/Akka.Discovery/Akka.Discovery.csproj
Comment (Aaronontheweb) [inline src/core/Akka.Discovery/Akka.Discovery.csproj:11]: need to add a tag here to mark this as a pre-release module

File: src/core/Akka.Discovery/Discovery.cs
Comment (Aaronontheweb) [inline src/core/Akka.Discovery/Discovery.cs:17]: Should probably have an XML-DOC comment here

**Review by Aaronontheweb — APPROVED** [review]
Some minor changes that I documented - but none worth blocking the merge of this PR. I'll send in a separate PR to add the beta tag for Akka.Discovery so we can include this as part of Akka.NET v1.4.12

**Comment by Aaronontheweb** [comment]
@ismaelhamed could you open an issue describing the Akka.IO DNS changes that we need to make in order to support DNS-based service discovery? We should start planning for those.

**Comment by mchandschuh** [comment]
Was DNS discovery ever implemented, and if not, is there an expected timeline and/or an issue available to track the progress (tried searching but didn't find it). Thanks!

**Pull request merged**
